### ISSUE TITLE
TypeCase broken on shapeless 2.3.3 only in scala 2.12.x

### ISSUE THREAD

**Issue by vascorsd** [milessabin/shapeless#848]
I was upgrading a current project to scala 2.12.6 and got into the following behaviour:

  * TypeCase compiles and works for shapeless version 2.3.2 on both scala 2.11.x and 2.12.x
  * TypeCase fails to compile using shapeless version 2.3.3 on scala 2.12.x but compiles as works fine on scala 2.11.x

The following error is thrown:

```
No default Typeable for type Protocol.SpecificCommandOne capturing an outer type variable
[error]   private val someMessage = TypeCase[Payloads.Payload[SpecificCommandOne]]
[error]                                     ^
[error] one error found
```

I managed to extract an example with a structure very similar to what I have on the project (which is much more complex) here - https://gitlab.com/vascorsd/broken-shapeless

Also made a scastie version here - https://scastie.scala-lang.org/6cODaCGGR62ZXsxoLiYfkg

They could probably be reduced much more, but that's what I got.

On the project I was only using shapeless 2.3.2 and didn't get that error until I also upgraded some dependencies and the shapeless version got forced to become the 2.3.3. Here the evicted log from sbt:

```
[info]     * com.chuusai:shapeless_2.12:2.3.3 is selected over 2.3.2
[info]         +- io.circe:circe-generic_2.12:0.9.3 ()               (depends on 2.3.3)
[info]         +- org.tpolecat:doobie-core_2.12:0.5.3                (depends on 2.3.3)
[info]         +- org.reactivemongo:reactivemongo_2.12:0.12.5 ()     (depends on 2.3.2)
```

Currently I solved the problem with the followng line on my sbt, forcing the shapeless version down as I really wanted to do the upgrade of the project to scala 2.12.6:

```
  dependencyOverrides ++= Seq(
    "com.chuusai" %% "shapeless" % "2.3.2"
  )
```

Hope this helps. I'm not sure if my current solution may break something else unexpectedly.

I read other similar bugs about on the bug tracker about TypeCase and Typeable but they don't seem to be this exact case.

**Comment by milessabin** [comment]
Interesting, thanks for the report.

It looks as though `TypeCase` is treating the nested `Prop` as if it were a captured type variable rather than as a stable concrete type ... if you move `Prop` to the top level your example compiles successfully.

I think it should work as written. Do you fancy taking a swing at fixing the bug?

**Comment by vascorsd** [comment]
Hi, thanks for confirming it's not a known bug.

I was looking before around the code I think is relevant to the issue in https://github.com/milessabin/shapeless/blob/808ba2384c257ff1d77e47522afb3ff3b78f7a81/core/src/main/scala/shapeless/typeable.scala#L454 and https://github.com/milessabin/shapeless/commit/808ba2384c257ff1d77e47522afb3ff3b78f7a81#diff-50e8c1d37a867a5cb03f58d644c92059 but I can't say I understand much of what the code is doing :/

Also as it seems dependent on the scala compiler version it is even more strange for me. 

Do you have any ideas or suggestions about what it could be?

I don't currently have much time to dedicate to problem at hand, so unless it is something really straightforward don't wait for me ;)

**Comment by milessabin** [comment]
It looks to me as though the problem is most likely in the definition of `closesOverType` just above.

**Comment by joroKr21** [comment]
I have a PR that touches this area #803. It removes the `closesOverType` method altogether. You could try if that works.

**Comment by joroKr21** [comment]
At least the discussion might be relevant, esp. the last comment https://github.com/milessabin/shapeless/pull/803#issuecomment-355751418

**Comment by ragb** [comment]
@joroKr21 seems to work properly with your PR... Although as this  example has no abstract type members or type projections, I still can't grasp what's going on... Mainy what differs from Scala 2.11 to 2.12 here. Still, I'm not confortable at all reasoning about   these sort of advanced macros and type system corner cases.

**Comment by vascorsd** [comment]
Hello again. 

I've got a little more time around this trying to figure out some kind of pattern and discovered a few more things. At least I got the example to compile under a specific case. I've updated the example code I first provided: https://gitlab.com/vascorsd/broken-shapeless

There's two different things going on here it seems: notice that in the `Main.scala` there are two `TypeCase` instances being generated. The first one compiles, the second one doesn't.

In the first one that compiles we got a top sealed trait called `Command`, then a `QueryCommand` trait that extends it and then an `InternalQuery` trait which then has case classes extending it. When we summon `TypeCase` for this middle trait `QueryCommand` it works. On the case that doesn't work it's just a `SpecificCommandOne` trait that immediately has the specific case classes extending it and when we summon the `TypeCase` it fails. So I believe this indirection may be related somehow.

I got the whole example to compile doing the following: on the `SpecificCommandOne` change the two vals to defs. It just works without further changes.

That's what I got. Hope it helps. Sorry for maybe being too confusing :/ .
